testsys: Support externally defined tests

ecpullen opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Closes #2516

**Description of changes:**

This pr adds the support for `TestType::Custom`. 

`CrdInput` now has a function to create a `BTreeMap` of values that can be templated in the yaml test file. Each `CrdCreator` also has the ability to provide additional fields that are _family_ specific. 

A new set of tables are also supported in `Test.toml`, `[<Specifier>.configuration.<Config-Name>]`. For a custom test named `smoke` for `aws-k8s` variants would have the following table header, `['aws-k8s'.configuration.smoke.x86_64]`. The arch is optional if the same configuration can be used across arches (`['aws-k8s'.configuration.smoke]`).

Custom tests can be run with `cargo make TESTSYS_TEST=<Name-of-test-in-Test.toml> test -f <Path-to-templated-yaml>`

I added information to `Testing.md` for custom test types, but decided not to add it to `Test.toml.example` to prevent it from becoming too complicated.

**Testing done:**

Tested with a custom test using the command above and ensured the test passed and the correct configuration was used.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

ecpullen committed:
testsys: Support external tests

Add support for the `Unknown` variant of the `TestType` enum. Templated yaml
files can be tested using `-f`.

webern commented on TESTING.md:
````diff
@@ -143,6 +143,26 @@ instance-type = "g5g.2xlarge"
 
+Tables can also be created for custom testing configurations. For a custom test type called `foo`, the config above can be updated: 
+
+```yaml
````
(Fix some of the others as well)

```suggestion
```toml
```

webern commented on TESTING.md:
```diff
@@ -302,3 +322,40 @@ cargo make -e TESTSYS_TEST=migration test
+
+First, a test agent needs to be constructed.
+The `test-agent-cli` provides an interface for creating bash based testing agents.
+Checkout the [runbook](https://github.com/bottlerocket-os/bottlerocket-test-system/blob/v0.0.3/agent/test-agent-cli/design/RUNBOOK.md) for instructions on creating an agent.
```
I think pointing to a specific version is brittle because people won't know to update it when we move to a newer version of testsys.

webern commented on tools/testsys-config/src/lib.rs:
```diff
@@ -148,6 +153,17 @@ impl GenericConfig {
+    /// Get the configuration for a specific test type.
+    pub fn test<S>(&self, test_type: S) -> GenericConfig
+    where
+        S: Into<String>,
```
```suggestion
        S: AsRef<str>,
```

webern commented on tools/testsys-config/src/lib.rs:
```diff
@@ -148,6 +153,17 @@ impl GenericConfig {
+        S: Into<String>,
+    {
+        self.configuration
+            .get(&test_type.into())
```
```suggestion
            .get(test_type.as_ref())
```

webern commented on tools/testsys/src/crds.rs:
```diff
@@ -127,8 +130,27 @@ impl<'a> CrdInput<'a> {
+/// Take the value of the `Option` or `"null"` if the `Option` was `None`
+fn some_or_null(field: &Option<String>) -> String {
+    field.to_owned().unwrap_or_else(|| "null".to_string())
+}
```
```suggestion
}

```

webern commented on tools/testsys/src/crds.rs:
```diff
@@ -309,10 +342,47 @@ pub(crate) trait CrdCreator: Sync {
-                        what: "Custom test types".to_string(),
-                    })
+                TestType::Unknown(test_type) => {
+                    let path = crd_template_file_path
+                        .as_ref()
+                        .context(error::InvalidSnafu {
+                            what: "A crd template file is required for custom test types.",
+                        })?;
```
This is bad. A function signature shouldn't allow this to happen.

webern commented on tools/testsys/src/crds.rs:
```diff
@@ -159,9 +181,20 @@ pub(crate) trait CrdCreator: Sync {
-    async fn create_crds(&self, test_type: TestType, crd_input: &CrdInput) -> Result<Vec<Crd>> {
+    async fn create_crds(
+        &self,
+        test_type: TestType,
+        crd_input: &CrdInput,
+        crd_template_file_path: Option<String>,
```
This function signature isn't ideal. You have an Enum, `test_type` and a field `crd_template_file_path` that is required only when `test_type` is `Unknown`.

Can you break up this function into, say, 2 functions, e.g. `create_crds` (which takes `KnownTestType`) and `create_custom_crds` which takes the name of the custom test and the `crd_template_file_path`?

This solves two problems, a quirky function signature, and a function definition that is already long. The caller would have to call either `create_crds` or `create_custom_crds`.

Incidentally, would `Custom` be a better name than `Unknown` throughout?

ecpullen commented:
^
* Renamed `TestType::Unknown` to `TestType::Custom`
* Corrected code block types
* Removed tag from `test-agent-cli` link
* Converted `Into<String>` to `AsRef<str>`
* Added extra line between functions
* Split up `create_crds` into 2 functions. One for custom and one for regular.

webern commented on tools/testsys/src/crds.rs:
```diff
@@ -159,167 +182,267 @@ pub(crate) trait CrdCreator: Sync {
     }
 }
 
+// TestType::Unknown(test_type) => {
```
delete commented code

ecpullen commented:
^ Remove commented code.

etungsten commented on TESTING.md:
```diff
@@ -302,3 +322,40 @@ cargo make -e TESTSYS_TEST=migration test
+
+First, a test agent needs to be constructed.
+The `test-agent-cli` provides an interface for creating bash based testing agents.
+Checkout the [runbook](https://github.com/bottlerocket-os/bottlerocket-test-system/blob/develop/agent/test-agent-cli/tests/data/deploy_test.yaml) for instructions on creating an agent.
```
Is this link suppose to point to here: https://github.com/bottlerocket-os/bottlerocket-test-system/blob/develop/agent/test-agent-cli/design/RUNBOOK.md?

ecpullen replied:
Yes. It has instructions for creating bash based testing agents. We can point to a different place if you prefer.

etungsten replied:
The current link is wrong then. It's currently pointing at https://github.com/bottlerocket-os/bottlerocket-test-system/blob/develop/agent/test-agent-cli/tests/data/deploy_test.yaml

ecpullen commented:
^ Fix link for `test-agent-cli`.
